### DISCUSSION TITLE
Any way to improve the convergence of ParIlut?

### DISCUSSION THREAD

**Discussion by FlyingCat2023** [ginkgo-project/ginkgo#1572] [Q&A]
I have a matrix which is generated from CutFEM. The matrix is symmetric but may be not positive definite. I try to solve Ax = b by using the ParIlut preconditioner but  fail to get a convergent solution. I found that it was because ILU was not converged in both 2D and 3D problems. I wonder if there is a solution to ensure convergence of ParIlut?

**Comment by pratikvn** [comment]
I believe ILU should be more robust than ParILU/ParILUT. If ILU does not converge, then I am not sure if there is a way to make ParILU/ParILUT converge.

**Reply by upsj** [reply]
If your problems are not too big, you could also try running ParILUT with the `ReferenceExecutor`, which will give you exact numerical calculations. If that one doesn't converge, you will also not have any luck with CUDA/HIP or OpenMP.

**Reply by nbeams** [reply]
I would also note that the convergence of ParILUT depends on the off-diagonal values of the matrix being < 1. @FlyingCat2023, if your matrix is mostly diagonally dominant, but the actual magnitude of off-diagonal values can be > 1, then scaling by the inverse of the diagonal of the matrix might make ParILUT converge. 

E.g. something like 
```
auto diag = mtx->extract_diagonal();
diag->inverse_apply(mtx, mtx);
```

**Reply by FlyingCat2023** [reply]
Hello all,

Sorry for late response. 
I have taken the actions as advised, as below are my feedback.
- With reference executor, ParIlut leads to convergence. Therefore, I concluded that the divergence is caused by the parallel algorithm
- With scaling as advised, OpenMPExecutor leads to convergence at the first iteration. However, for further iterations, it converges to some extent, but not meeting the tolerance requirements, e.g., 1e-5.
I would like to hear more from you.

**Reply by nbeams** [reply]
Hi @FlyingCat2023,

How many iterations are you doing? Would you mind sharing your convergence history for each iteration?

After the scaling, is every single off-diagonal value < 1? Or just almost all of them? If you print out the final L and U factors from the sequential (reference) algorithm, are all off-diagonal values < 1? If not, we can't guarantee that the fixed-point iteration will converge "cleanly". There is some discussion about this in the 2015 ParILU paper by Chow and Patel (SIAM J. Sci. Comput. Vol. 37, No. 2) if you want to read more.

However, another thing you can try is using the factorization in the preconditioner anyway, even if it hasn't converged. The residual that it is reporting is based on how well the factorization satisfies the condition that A-LU = 0 on the sparsity pattern of A. Just because a factorization doesn't satisfy this exactly, doesn't necessarily mean it won't work well in a preconditioner. I have personally run into this with a matrix, where the ILU residual was still large (larger than 1e-5, even) after several iterations, yet the preconditioner produced with this factorization worked nearly as well (in terms of iterations required to solve) as the sequential ILU, for which the ILU residual was 1e-12. So it may be worth a try for your problem.

**Reply by FlyingCat2023** [reply]
Hello, I am late again. Unfortunately, The off-diagonal value are not < 1 by scaling them with the diagonal elements. I found some diagonal elements are very small as compared to the off-diagonal elements around it.
Here are the results.
parilu 2 2 2
Reading ../data/a_1_0.mtx
1;1689553197;1638503874;0;0;
2;3315644637;3236100911;0;0;
parilu 2 2 2
Reading ../data/a_1_0.mtx
1;155339553;192793444;6.52954e-08;6.52954e-08;
2;275994243;296592474;6.52954e-08;6.52954e-08;
parilut 2 2 2
Reading ../data/a_1_0.mtx
1;3837950080;3671037208;6.52954e-08;6.52954e-08;
2;21346279645;22399902701;0.125565;0.125565;
parilut 4 2 2
Reading ../data/a_1_0.mtx
1;3826550646;3740850592;6.52954e-08;6.52954e-08;
2;22383866145;22235150445;0.125565;0.125565;
3;49282028763;51056535073;135.172;513.352;
4;79245004217;79275728756;1325.01;1331.49;

**Reply by nbeams** [reply]
> I found some diagonal elements are very small as compared to the off-diagonal elements around it.

This will definitely cause problems for ParILU(T). You could try a reordering meant to move large values to the diagonal. Ginkgo has the Mc64 reordering algorithm in its `experimental` namespace: https://ginkgo-project.github.io/ginkgo-generated-documentation/doc/develop/classgko_1_1experimental_1_1reorder_1_1Mc64.html

An example of how you could use it:

```
      // Create the Mc64 object
       auto perm_comp = gko::experimental::reorder::Mc64<ValueType,IndexType>::build()
                   .on(exec)->generate(orig_mtx);
       // Get row/column operators
       auto row_perm_scale = gko::as<gko::matrix::ScaledPermutation<ValueType,IndexType>>(
                      perm_comp->get_operators()[0]);
       auto col_perm_scale = gko::as<gko::matrix::ScaledPermutation<ValueType,IndexType>>(
                      perm_comp->get_operators()[1]);
       // Make a new matrix with Mc64 applied
       mtx_op = gko::share(orig_mtx->scale_permute(row_perm_scale, col_perm_scale));
```

Which Executor were you using for the results you added? Previously you said "OpenMPExecutor leads to convergence at the first iteration. However, for further iterations, it converges to some extent, but not meeting the tolerance requirements, e.g., 1e-5," which is not what these results show. I also find it very odd for the residual to be 0 after 1 iteration in the first test...

**Reply by FlyingCat2023** [reply]
Sorry for the confusion. 
The first example is using the Reference Executor. The others are using OMP Executor.

**Reply by FlyingCat2023** [reply]
Test 2 converge to tolerance by using ParILU. However, Test 3 and 4 does not converges by using ParILUT, starting from iteration 3, as can be seen in test 4.

**Reply by nbeams** [reply]
But the residuals for 1 or 2 iterations are the same in test 2. It's very odd for ParILU to converge after only 1 iteration.
Regardless, if the matrix is not diagonally dominant, then we can't expect that it will converge quickly. We could get lucky -- but we can't guarantee that it should be converging. A reordering may help, or else ParILU(T) may not work for your matrix.
